### PR TITLE
Minor speed-ups by not constructing lists for all and any and using tuples for constant iterators

### PR DESCRIPTION
**Pull request by oscargus**
#### Brief description of what is fixed or changed
Not creating an explicit list is faster for `any` and `all`, which can terminate early if an iterator is provided instead.

When creating an iterator with constants, tuples tend to be faster than lists.

#### Other comments
Some minor modifications to e.g. avoid computing lengths several times.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/categories/diagram_drawing.py [hunk 1]
```diff
@@ -2155,14 +2155,14 @@ def _check_free_space_vertical(dom_i, cod_i, dom_j, grid):
...
+            free_left = all(not grid[i, dom_j - 1] for i in
+                            range(start, end + 1))
 
         if dom_j == grid.width - 1:
             free_right = True
         else:
-            free_right = all([not grid[i, dom_j + 1] for i in
-                              range(start, end + 1)])
+            free_right = all(not grid[i, dom_j + 1] for i in
+                             range(start, end + 1))
```
Comment (oscarbenjamin) [inline sympy/categories/diagram_drawing.py:2165]: I suppose it is probably faster to do `not any` rather than `all not` because only one negation is needed.

File: sympy/combinatorics/homomorphisms.py [hunk 2]
```diff
@@ -301,8 +301,8 @@ def homomorphism(domain, codomain, gens, images=[], check=True):
     gens = list(gens)
     images = list(images)
 
-    images.extend([codomain.identity]*(len(generators)-len(images)))
-    gens.extend([g for g in generators if g not in gens])
+    images.extend((codomain.identity,) * (len(generators)-len(images)))
+    gens.extend(g for g in generators if g not in gens)
```
Comment (oscarbenjamin) [inline sympy/combinatorics/homomorphisms.py:305]: I'm not sure the second change has a speed improvement

Comment (oscarbenjamin) [reply sympy/combinatorics/homomorphisms.py:305]: Also if `gens` can be a large list then `g not in gens` is an `O(N)` containment test whereas a set could give `O(1)` containment testing. Potentially that makes the difference between the operation as a whole being `O(N)` or `O(N^2)`.

File: sympy/combinatorics/perm_groups.py [hunk 3]
```diff
@@ -3679,7 +3679,7 @@ def schreier_sims_incremental(self, base=None, gens=None, slp_dict=False):
                 strong_gens_slp[g] = [g]
             return (_base, strong_gens, strong_gens_slp)
 
-        strong_gens.extend([k for k, _ in strong_gens_slp])
+        strong_gens.extend(k for k, _ in strong_gens_slp)
```
Comment (oscarbenjamin) [inline sympy/combinatorics/perm_groups.py:3682]: Not sure this is an improvement

File: sympy/categories/diagram_drawing.py (hunk 1 above)
Comment (oscargus) [reply sympy/categories/diagram_drawing.py:2165]: You are probably correct. I have not considered that type of changes, but can have a quick look.

File: sympy/combinatorics/permutations.py [hunk 4]
```diff
@@ -1153,7 +1153,7 @@ def full_cyclic_form(self):
         """
         need = set(range(self.size)) - set(flatten(self.cyclic_form))
         rv = self.cyclic_form
-        rv.extend([[i] for i in need])
+        rv.extend([i] for i in need)
```
Comment (oscarbenjamin) [inline sympy/combinatorics/permutations.py:1156]: Maybe
```
rv = self.cyclic_form + [[i] for i in need]
```

File: sympy/combinatorics/tensor_can.py [hunk 5]
```diff
@@ -1141,11 +1141,11 @@ def _get_bsgs(G, base, gens, free_indices):
         a.extend(ind2)
         a.extend(ind1)
         base_comm.append(ind1[0])
-        a.extend(list(range(ind2[-1] + 1, nr)))
+        a.extend(range(ind2[-1] + 1, nr))
         if sym == 0:
-            a.extend([nr, nr + 1])
+            a.extend((nr, nr + 1))
```
Comment (oscarbenjamin) [inline sympy/combinatorics/tensor_can.py:1146]: Is this sort of thing actually faster?

I get that if the tuple elements are compile-time constants then the tuple can be pre-created and marshalled into the pyc file but if the tuple elements depend on local variables as is the case here then we're just talking about the difference between the time to construct a tuple vs a list:
```python
def f1(l, a):
    for _ in range(100):
        l.extend([a, a+1])

def f2(l, a):
    for _ in range(100):
        l.extend((a, a+1))

def f3(l, a):
    for _ in range(100):
        l += [a, a+1]

def f4(l, a):
    for _ in range(100):
        l.append(a)
        l.append(a+1)
```
Timings:
```python
In [28]: l = []

In [29]: %timeit f1(l, 5)
30 µs ± 105 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [30]: l = []

In [31]: %timeit f2(l, 5)
26.3 µs ± 107 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [32]: l = []

In [33]: %timeit f3(l, 5)
23.1 µs ± 2.55 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [34]: l = []

In [35]: %timeit f4(l, 5)
31.5 µs ± 244 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
Looks like `+=` with a list is fastest.

File: sympy/geometry/entity.py [hunk 6]
```diff
@@ -171,7 +171,7 @@ def _repr_svg_(self):
             # will fall back to the next representation
             return None
 
-        if any([not x.is_number or not x.is_finite for x in bounds]):
+        if any(not x.is_number or not x.is_finite for x in bounds):
```
Comment (oscarbenjamin) [inline sympy/geometry/entity.py:174]: Might be clearer and faster as
```python
if not all(x.is_number and x.is_finite for x in bounds):
```
More importantly it might be more correct. What should happen here if `x.is_finite` gives `None`?

File: sympy/physics/continuum_mechanics/beam.py [hunk 7]
```diff
@@ -901,8 +901,8 @@ def max_shear_force(self):
                 val = []
                 for point in points:
                     val.append(shear_curve.subs(x, point))
-                points.extend([singularity[i-1], s])
-                val.extend([limit(shear_curve, x, singularity[i-1], '+'), limit(shear_curve, x, s, '-')])
+                points.extend((singularity[i-1], s))
+                val.extend((limit(shear_curve, x, singularity[i-1], '+'), limit(shear_curve, x, s, '-')))
                 val = list(map(abs, val))
                 max_shear = max(val)
```
Comment (oscarbenjamin) [inline sympy/physics/continuum_mechanics/beam.py:907]: Maybe
```python
val += [
    abs(limit(shear_curve, x, singularity[i-1], '+')),
    abs(limit(shear_curve, x, s, '-'))),
]
```
I guess this isn't tested with symbolic inputs because `max(val)` would probably fail in that case.

File: sympy/solvers/ode/hypergeometric.py [hunk 8]
```diff
@@ -46,7 +46,7 @@ def match_2nd_hypergeometric(eq, func):
     r = collect(eq,
         [func.diff(x, 2), func.diff(x), func]).match(deq)
     if r:
-        if not all([r[key].is_polynomial() for key in r]):
+        if not all(r[key].is_polynomial() for key in r):
```
Comment (oscarbenjamin) [inline sympy/solvers/ode/hypergeometric.py:49]: Maybe
```python
if not all(val.is_polynomial() for val in r.values()):
```

File: sympy/stats/frv_types.py [hunk 9]
```diff
@@ -785,7 +785,7 @@ def set(self):
 
     @property
     def is_symbolic(self):
-        return not all([self.k.is_number, self.c.is_number, self.delta.is_number])
+        return not all((self.k.is_number, self.c.is_number, self.delta.is_number))
```
Comment (oscarbenjamin) [inline sympy/stats/frv_types.py:788]:
```python
return not (self.k.is_number and self.c.is_number and self.delta.is_number)
```
or perhaps
```python
return not all(s.is_number for s in (self.k, self.c, self.delta))
```
Not sure if classes in stats have something like `self.parameters` for this.

File: sympy/stats/tests/test_rv.py [hunk 10]
```diff
@@ -214,12 +214,11 @@ def test_Sample():
     if not numpy:
         skip('Numpy is not installed. Abort tests')
     #Test Issue #21563: Output of sample must be a float or array
-    assert isinstance(sample(X), numpy.int64)
+    assert isinstance(sample(X), (numpy.int32, numpy.int64))
```
Comment (oscarbenjamin) [inline sympy/stats/tests/test_rv.py:217]: Why is this changed?

File: sympy/tensor/index_methods.py [hunk 11]
```diff
@@ -188,7 +188,7 @@ def _get_indices_Add(expr):
     if not non_scalars:
         return set(), {}
 
-    if not all([x == non_scalars[0] for x in non_scalars[1:]]):
+    if not all(x == non_scalars[0] for x in non_scalars[1:]):
```
Comment (oscarbenjamin) [inline sympy/tensor/index_methods.py:191]: Maybe
```
if len(set(non_scalars)) > 1:
```

**Comment by oscarbenjamin** [comment]
The changes to make shortcutting possible are definitely good e.g.
```python
all([x for x in e]) -> all(x for x in e)
```

I'm not sure about changes like
```python
a.extend([y, z]) -> a.extend((y, z))
```
partly just because calling `extend` with a tuple looks weird to me. The fastest version of this according to my timings above is
```python
a += [y, z]
```
which I also think looks nicer and clearer.

File: sympy/combinatorics/homomorphisms.py (hunk 2 above)
Comment (oscargus) [reply sympy/combinatorics/homomorphisms.py:305]: You are probably correct. I'll revert those (and see if I find other where a list may help).

File: sympy/combinatorics/tensor_can.py (hunk 5 above)
Comment (oscargus) [reply sympy/combinatorics/tensor_can.py:1146]: I didn't consider adding to the list as an option, but it indeed looks faster. I read https://stackoverflow.com/questions/14446128/python-append-vs-extend-efficiency where no one seem to have thought about it either.

File: sympy/geometry/entity.py (hunk 6 above)
Comment (oscargus) [reply sympy/geometry/entity.py:174]: I do not analyze the code to that extent... :-) But I do indeed see your point!

File: sympy/physics/continuum_mechanics/beam.py (hunk 7 above)
Comment (oscargus) [reply sympy/physics/continuum_mechanics/beam.py:907]: Yes, plus `val.append(abs(shear_curve.subs(x, point)))` on the lines above.

File: sympy/stats/tests/test_rv.py (hunk 10 above)
Comment (oscargus) [reply sympy/stats/tests/test_rv.py:217]: For some reason it returned `numpy.int32` on my computer. Not sure why though. I do not think it relates to the changes I've made, so I guess that it may return that on 32-bit machines/numpy/python (which I do not think I have...).

File: sympy/combinatorics/permutations.py [hunk 12]
```diff
@@ -983,7 +983,7 @@ def __new__(cls, *args, size=None, **kwargs):
         temp = set(temp)
 
         if not is_cycle:
-            if any(i not in temp for i in range(len(temp))):
```
Comment (oscargus) [inline sympy/combinatorics/permutations.py:986]: Not sure if this actually helps performance, but tends to be more readable in my opinion. It may avoid multiple not, but not sure if `not in` is one or two statements.

File: sympy/combinatorics/tensor_can.py (hunk 5 above)
Comment (oscargus) [reply sympy/combinatorics/tensor_can.py:1146]: I guess there is a threshold for when `+=` is faster? Either way, for most cases I kept the tuples and extended.

File: sympy/categories/diagram_drawing.py [hunk 13]
```diff
@@ -2155,14 +2155,14 @@ def _check_free_space_vertical(dom_i, cod_i, dom_j, grid):
         if dom_j == 0:
             free_left = True
         else:
-            free_left = all([not grid[i, dom_j - 1] for i in
-                             range(start, end + 1)])
+            free_left = not all(grid[i, dom_j - 1] for i in
+                                range(start, end + 1))
```
Comment (oscarbenjamin) [inline sympy/categories/diagram_drawing.py:2159]: Should this not be changed to `any`?

File: sympy/combinatorics/fp_groups.py [hunk 14]
```diff
@@ -259,7 +259,7 @@ def _is_infinite(self):
         used_gens = set()
         for r in self.relators:
             used_gens.update(r.contains_generators())
-        if any([g not in used_gens for g in self.generators]):
+        if not all(g in used_gens for g in self.generators):
```
Comment (oscarbenjamin) [inline sympy/combinatorics/fp_groups.py:262]: I'm not sure what the type of `self.generators` is but `used_gens` is a set and this looks a subset test to me. The most efficient way would be:
```python
if not set(self.generators) <= used_gens:
```
It's not possible to negate the `<=` because this is a partial order.

File: sympy/solvers/ode/hypergeometric.py (hunk 8 above)
Comment (oscargus) [reply sympy/solvers/ode/hypergeometric.py:49]: One should try to figure out a way to detect this in the code...

File: sympy/stats/tests/test_rv.py (hunk 10 above)
Comment (oscargus) [reply sympy/stats/tests/test_rv.py:217]: And any integer type should be OK considering that there are 6 options...

File: sympy/categories/diagram_drawing.py (hunk 13 above)
Comment (oscargus) [reply sympy/categories/diagram_drawing.py:2159]: It should!

I'll also try to revert the changes fully in files where there are now only white space changes...

File: sympy/sets/handlers/intersection.py [hunk 15]
```diff
@@ -264,7 +264,7 @@ def intersection_sets(self, other): # noqa:F811
             # not helpful here: multiple parametric solutions.
             if len(solns) == 0:
                 return EmptySet
-            elif any(not isinstance(s, int) and s.free_symbols
+            elif any(s.free_symbols and not isinstance(s, int)
```
Comment (oscarbenjamin) [inline sympy/sets/handlers/intersection.py:267]: It is possible for `s` to be an `int` here? If it is then `s.free_symbols` would give an attribute error.

My guess is that this is a workaround for a bug that resulted in unsympified types being returned from `diophantine`. I think the `not isinstance(s, int)` should be removed if the tests pass with this change. If there is a failure then I guess that's a bug in diophantine. It should be fixed rather than worked around.

Comment (oscargus) [reply sympy/sets/handlers/intersection.py:267]: You are probably correct. I realized that these changes were not always good when I found the `x.is_Integer and ...`, but forgot about this one. Will not fix diophantine though. :-)

Comment (oscargus) [reply sympy/sets/handlers/intersection.py:267]: Seems like it passes. Should we assume that the bug is fixed or that a proper test case is missing?

File: sympy/tensor/array/expressions/array_expressions.py [hunk 16]
```diff
@@ -44,7 +44,7 @@ def __getitem__(self, item):
         return ArrayElement(self, item)
 
     def as_explicit(self):
-        if any(not isinstance(i, (int, Integer)) for i in self.shape):
+        if not all(isinstance(i, (int, Integer)) for i in self.shape):
```
Comment (oscarbenjamin) [inline sympy/tensor/array/expressions/array_expressions.py:47]: This looks like something to be fixed elsewhere. The shape should really be required to be sympified or required to be an `int`. Allowing both is flakey.

File: sympy/sets/handlers/intersection.py (hunk 15 above)
Comment (oscargus) [reply sympy/sets/handlers/intersection.py:267]: Or rather the diophantine tests passed. Better wait for the other ones as well...

File: sympy/combinatorics/permutations.py (hunk 12 above)
Comment (oscarbenjamin) [reply sympy/combinatorics/permutations.py:986]: Since `temp` is a set here this can be something like:
```python
if temp != set(range(len(temp))):
```

File: sympy/sets/handlers/intersection.py (hunk 15 above)
Comment (oscarbenjamin) [reply sympy/sets/handlers/intersection.py:267]: Something like this is designed to hide bugs. I'd rather remove it so that the bugs can be seen and fixed even if that has to happen in future rather than now.

File: sympy/categories/diagram_drawing.py [hunk 17]
```diff
@@ -670,7 +670,7 @@ def _weld_triangle(tri, welding_edge, fringe, grid, skeleton):
         a = (a[0] + offset[0], a[1] + offset[1])
         b = (b[0] + offset[0], b[1] + offset[1])
 
-        fringe.extend([(a, target_cell), (b, target_cell)])
+        fringe.extend(((a, target_cell), (b, target_cell)))
```
Comment (oscarbenjamin) [inline sympy/categories/diagram_drawing.py:673]: There are still quite a few changes like this. I think this would be better left unchanged or should be changed to
```python
fringe += [(a, target_cell), (b, target_cell)]
```
since that's the fastest version.

Comment (oscargus) [reply sympy/categories/diagram_drawing.py:673]: Now, there are a few aspects here, just to be clear:
* If the content is compile-time constant?
* More than two terms?
* Some compile-time constants?

File: sympy/combinatorics/homomorphisms.py [hunk 18]
```diff
@@ -301,7 +301,7 @@ def homomorphism(domain, codomain, gens, images=[], check=True):
     gens = list(gens)
     images = list(images)
 
-    images.extend([codomain.identity]*(len(generators)-len(images)))
+    images.extend((codomain.identity,) * (len(generators)-len(images)))
```
Comment (oscarbenjamin) [inline sympy/combinatorics/homomorphisms.py:304]: What advantage does a change like this give?

It could alternatively be:
```python
images += [codomain.identity] * (len(generators)-len(images))
```
I think that's potentially fastest and also nicer to read IMO.

Comment (oscarbenjamin) [reply sympy/combinatorics/homomorphisms.py:304]: There are many other changes like this one

Comment (oscargus) [reply sympy/combinatorics/homomorphisms.py:304]: Speed, I'd say. But your argument is primarily against variables not constants?

Not sure one can say that += will always be faster?

File: sympy/combinatorics/tensor_can.py (hunk 5 above)
Comment (oscarbenjamin) [reply sympy/combinatorics/tensor_can.py:1146]: I think that `+=` with a list is faster when the list of new items is small (e.g. length 2 here). The reason is just that in CPython an operator like `+` is more efficient than a function/method call for e.g. `extend`.

File: sympy/categories/diagram_drawing.py (hunk 17 above)
Comment (oscarbenjamin) [reply sympy/categories/diagram_drawing.py:673]: If the tuple is entirely a compile time constant then a tuple expression like `(1, 2)` is efficient because it is *fully* precomputed. Otherwise I don't think tuples have any advantage in this situation. Creating a small tuple from a literal tuple expression should be just as fast as creating a small list and both should be just as fast when iterating.

Using `+=` will usually be faster than `extend` because operators are faster than methods in CPython. This will only be significant when extending by a small list.

Using a generator with `extend` is slower on a per-item basis because both `list.extend` and `list.__iadd__` have fast paths for concrete containers like list and tuple that can not be used with a generator.

File: sympy/combinatorics/homomorphisms.py (hunk 18 above)
Comment (oscarbenjamin) [reply sympy/combinatorics/homomorphisms.py:304]: Can you show a timing that demonstrates the speed improvement?

Comment (oscarbenjamin) [reply sympy/combinatorics/homomorphisms.py:304]: Given these examples:
```python
def f1(a):
    l = []
    for _ in range(100):
        l.extend([a] * 100)

def f2(a):
    l = []
    for _ in range(100):
        l.extend((a,) * 100)

def f3(a):
    l = []
    for _ in range(100):
        l += [a] * 100
```
I get the timings
```python
In [22]: %timeit f1(5)
210 µs ± 8.98 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [23]: %timeit f2(5)
248 µs ± 2.79 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [24]: %timeit f3(5)
213 µs ± 11.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
According to that `f2` (using a tuple) is slowest. I'm not sure how well that generalises but it doesn't support making the change here.

Ultimately I think that if we're not sure about a change it's best just to stick with what we have.

File: sympy/combinatorics/homomorphisms.py (hunk 2 above)
Comment (smichr) [reply sympy/combinatorics/homomorphisms.py]: What about `gens = list(uniq(gens + generators))`?

File: sympy/stats/tests/test_rv.py (hunk 10 above)
Comment (oscargus) [reply sympy/stats/tests/test_rv.py:217]: I can just confirm that I get this on my current master as well. Must have installed a 32-bit version by mistake?

Anyway, I think it makes sense that this passes even if you have a 32-bit version locally. (Noted that I was running pywin32 although I did install the 64-bit miniconda as far as I (and my web history) can tell...)

File: sympy/categories/diagram_drawing.py (hunk 17 above)
Comment (oscargus) [reply sympy/categories/diagram_drawing.py:673]: One could argue that creation of tuples is slightly faster, at least in terms of "python operations", and I've seen minor benefits even then. I guess there is a reason why the points are tuples and not lists. (Even though one can also argue that they should be tuples as they do not need to be manipulated.)

I'll get rid of the generator extends I've added, but here I do actually only see benefits (unless we should change all `extend` to `+=` which I will not do currently).

File: sympy/combinatorics/homomorphisms.py (hunk 18 above)
Comment (oscargus) [reply sympy/combinatorics/homomorphisms.py:304]: OK! I might have generalized things a bit to far. With constants there is a huge speed-up using tuples, but not using variables.

Comment (oscargus) [reply sympy/combinatorics/homomorphisms.py:304]:
```
%timeit (x,)*100
621 ns ± 3.82 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit [x,]*100
578 ns ± 7.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit [1,]*100
562 ns ± 29.7 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit (1,)*100
14.6 ns ± 0.25 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
```

File: sympy/stats/tests/test_rv.py (hunk 10 above)
Comment (oscarbenjamin) [reply sympy/stats/tests/test_rv.py:217]: Yes, I think this is fine.

File: sympy/combinatorics/homomorphisms.py (hunk 18 above)
Comment (oscarbenjamin) [reply sympy/combinatorics/homomorphisms.py:304]: I think that the huge speedup is only when the whole tuple can be precomputed at compile time.

Otherwise there is no reason in principle why creating a tuple should be faster than creating a list either for small or large sizes.

File: sympy/tensor/index_methods.py (hunk 11 above)
Comment (oscargus) [reply sympy/tensor/index_methods.py:191]:
```
    if len(set(non_scalars)) > 1:
TypeError: unhashable type: 'set'
```

**Comment by oscargus** [comment]
Now I think I've reverted all the extends.

I only changed to tuples where there were constants only.

If it is OK now, I will squash the commits before merging.

File: sympy/geometry/entity.py [hunk 19]
```diff
@@ -171,7 +171,7 @@ def _repr_svg_(self):
             # will fall back to the next representation
             return None
 
-        if any([not x.is_number or not x.is_finite for x in bounds]):
+        if not all(x.is_number and x.is_finite for x in bounds):
             return None
```
Comment (oscarbenjamin) [inline sympy/geometry/entity.py:175]: Does this change have an observable effect in the case that `is_finite` is None?

Maybe there could be a test for that...

Comment (oscarbenjamin) [inline sympy/geometry/entity.py:175]: Does this change have an observable effect in the case that `is_finite` is None?

Maybe there could be a test for that...

File: sympy/sets/sets.py [hunk 20]
```diff
@@ -867,7 +867,7 @@ def __len__(self):
         return reduce(lambda a, b: a*b, (len(s) for s in self.args))
 
     def __bool__(self):
-        return all([bool(s) for s in self.sets])
+        return all(bool(s) for s in self.sets)
```
Comment (oscarbenjamin) [inline sympy/sets/sets.py:870]: Is `bool` needed inside `all`?

Comment (oscarbenjamin) [reply sympy/sets/sets.py:870]: I would have thought this could just be `all(self.sets)`.

**Comment by oscarbenjamin** [comment]
I left some comments but they are not significant.

I think this looks good. You said you wanted to squash so go ahead and squash and merge.

Is there a good reason for squashing? Is it because there are changes that were reverted?

File: sympy/geometry/entity.py (hunk 19 above)
Comment (oscargus) [reply sympy/geometry/entity.py:175]: It may indeed be, but since this is printing code, I would assume that we will get some feedback pretty quickly... Right now, I do not really understand the code well enough to try to test it.

File: sympy/tensor/array/expressions/array_expressions.py (hunk 16 above)
Comment (oscargus) [reply sympy/tensor/array/expressions/array_expressions.py:48]: Indeed! I will not change it here though.

File: sympy/sets/sets.py (hunk 20 above)
Comment (oscargus) [reply sympy/sets/sets.py]: Thanks! Changed it and it seems to work.

**Comment by oscargus** [comment]
> Is it because there are changes that were reverted?

Yes. Thinking it is a hassle if this leads to merge conflicts somewhere else because of reverted changes.

**Comment by oscargus** [comment]
Fixed the `bool` thing and some minor things pointed out by pylint in the sets file (duplicate imports and using `set` as a variable name) as well and rebased. Hopefully this is good to go now.

File: sympy/combinatorics/tensor_can.py [hunk 21]
```diff
@@ -763,12 +763,12 @@ def canonicalize(g, dummies, msym, *v):
     """
     from sympy.combinatorics.testutil import canonicalize_naive
     if not isinstance(msym, list):
-        if not msym in [0, 1, None]:
+        if not msym in (0, 1, None):
```
Comment (oscarbenjamin) [inline sympy/combinatorics/tensor_can.py:766]: Not really important but I think `x not in y` is generally better for clarity then `not x in y` and also it's very slightly faster.

**Comment by oscarbenjamin** [comment]
Thanks
